### DISCUSSION TITLE
understanding sliding window batch size parameter ( sw_batch_size )?

### DISCUSSION THREAD

**Discussion by neuronflow** [Project-MONAI/MONAI#2866] [Q&A]
https://docs.monai.io/en/latest/_modules/monai/inferers/utils.html#sliding_window_inference
What is the intended use case for the batch size parameter for the sliding window inferer, `sw_batch_size`?

During my tests on various datasets, I always had best results when prioritizing window_size over batch_size.

How are the batches sampled?

I thought multiple batches from various locations in the volume might help to keep normalization etc. on a reasonable level, however during my tests I could not onserve this.

https://docs.monai.io/en/latest/_modules/monai/inferers/utils.html#sliding_window_inference

**Comment by Nic-Ma** [comment]
Hi @neuronflow ,

The `sw_batch_size` is designed to run multiple windows on GPU together for better performance(speed).
For example, maybe the network was trained with (96, 96, 96) input data size, to inference 1 (512, 512, 512) image, we can use batch_size to accelerate (96, 96, 96) windows.

Thanks.

**Comment by neuronflow** [comment]
@Nic-Ma thanks this is how I also understood it. How are the batches sampled from the (512, 512, 512) volume?

During my limited testing I found much better performance doing inference with (192,192,96) and a `sw_batch_size` of `1`  instead of (96,96,96) and a `sw_batch_size` of `4`.

(in practice I used a bit different window sizes)

I believe in my case this is mostly driven by normalization. The small samples from the edges sometimes contained only background and than the normalization would "amplify" the background signal.

Hence I am asking how they are sampled?

**Reply by Nic-Ma** [reply]
Hi @neuronflow ,

Here is the logic to generate windows based on window size and image size:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/inferers/utils.py#L111

Thanks.

**Comment by neuronflow** [comment]
Thank you, I am not sure if I completely understand the code.

A practical example to better illustrate my question:

I trained a network on  patches (192,192,64) with batch size `8` on large resolution microscopy images.

Now I want use the network to predict a single volume approx. (6000,4000,3000).
I usually try to fully exploit the available GPU RAM, so I increase `window_size` and `sw_batch_size` to the maximum. 

Do you have a recommendation how to best balance `window_size` and `sw_batch_size`?
During my small experiments I always had better results when prioritizing `window_size`.

**Reply by Nic-Ma** [reply]
Hi @neuronflow ,

That's an interesting question.
I think it depends on the dataset and network, some networks may not work well if the inference window is different from the train window. And some dataset content is sensitive to the window size.
@wyli @ericspod Do you guys have other recommendations here?

Thanks in advance.
